File: engine/src/main/java/com/terralite/engine/input/MovementIntent.java
package com.terralite.engine.input;

public record MovementIntent(double x, double y, double z) {
    public static final MovementIntent NONE = new MovementIntent(0.0, 0.0, 0.0);

    public static MovementIntent from(InputState input) {
        double x = 0.0;
        double z = 0.0;

        if (input.isPressed(InputActions.MOVE_LEFT)) {
            x -= 1.0;
        }
        if (input.isPressed(InputActions.MOVE_RIGHT)) {
            x += 1.0;
        }
        if (input.isPressed(InputActions.MOVE_FORWARD)) {
            z += 1.0;
        }
        if (input.isPressed(InputActions.MOVE_BACK)) {
            z -= 1.0;
        }

        return new MovementIntent(x, input.isPressed(InputActions.JUMP) ? 1.0 : 0.0, z);
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/input/MovementIntent.kt
package com.terralite.engine.input

@JvmRecord
data class MovementIntent(val x: Double, val y: Double, val z: Double) {
    companion object {
        @JvmField val NONE = MovementIntent(0.0, 0.0, 0.0)

        @JvmStatic fun from(input: InputState): MovementIntent {
            var x = 0.0; var z = 0.0
            if (input.isPressed(InputActions.MOVE_LEFT))    x -= 1.0
            if (input.isPressed(InputActions.MOVE_RIGHT))   x += 1.0
            if (input.isPressed(InputActions.MOVE_FORWARD)) z += 1.0
            if (input.isPressed(InputActions.MOVE_BACK))    z -= 1.0
            return MovementIntent(x, if (input.isPressed(InputActions.JUMP)) 1.0 else 0.0, z)
        }
    }
}
